src/librustc_trans/mir/block.rs

```diff
@@ -685,7 +687,37 @@ impl<'bcx, 'tcx> MirContext<'bcx, 'tcx> {
 
         match dest {
             Nothing => (),
-            Store(dst) => ret_ty.store(bcx, op.immediate(), dst),
+            Store(dst) => {
+                if let Some(llcast_ty) = ret_ty.cast {
+                    let ccx = bcx.ccx();
+                    // The actual return type is a struct, but the ABI
+                    // adaption code has cast it into some scalar type.
+                    // We can't use `ret_ty.store(..)` since that simply
+                    // does a `PointerCast` and a `Store`. Unfortunately,
+                    // that may run afoul of strict aliasing rules, leading
+                    // to wacky results after optimization since LLVM
+                    // may just replace it with an `undef`.
```

Comment:
The problem has nothing to do with aliasing rules... the problem is the buffer overflow.